pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1870 @@
+                return (None, 0u64.into());
+            }
+
+            let mint = |amt: BalanceOf<T>, to: T::AccountId| {
+                if let Ok(amount_transferred) =
+                    <T as Config>::Currency::deposit_into_existing(&to, amt)
+                {
+                    Self::deposit_event(Event::Rewarded {
+                        account: to.clone(),
+                        rewards: amount_transferred.peek(),
+                    });
+                }
+            };
+
+            let collator_fee = payout_info.collator_commission;
+            let collator_issuance = collator_fee * payout_info.round_issuance;
+
+            if let Some((collator, pts)) =
+                <AwardedPts<T>>::iter_prefix(paid_for_round).drain().next()
+            {
+                let mut extra_weight = 0;
+                let pct_due = Perbill::from_rational(pts, total_points);
+                let total_paid = pct_due * payout_info.total_staking_reward;
+                let mut amt_due = total_paid;
+                // Take the snapshot of block author and delegations
+                let state = <AtStake<T>>::take(paid_for_round, &collator);
+                let num_delegators = state.delegations.len();
+                if state.delegations.is_empty() {
+                    // solo collator with no delegators
+                    mint(amt_due, collator.clone());
+                    extra_weight += T::OnCollatorPayout::on_collator_payout(
+                        paid_for_round,
+                        collator.clone(),
+                        amt_due,
+                    );
+                } else {
+                    // pay collator first; commission + due_portion
+                    let collator_pct = Perbill::from_rational(state.bond, state.total);
+                    let commission = pct_due * collator_issuance;
+                    amt_due = amt_due.saturating_sub(commission);
+                    let collator_reward = (collator_pct * amt_due).saturating_add(commission);
+                    mint(collator_reward, collator.clone());
+                    extra_weight += T::OnCollatorPayout::on_collator_payout(
+                        paid_for_round,
+                        collator.clone(),
+                        collator_reward,
+                    );
+                    // pay delegators due portion
+                    for Bond { owner, amount } in state.delegations {
+                        let percent = Perbill::from_rational(amount, state.total);
+                        let due = percent * amt_due;
+                        if !due.is_zero() {
+                            mint(due, owner.clone());
+                        }
+                    }
+                }
+
+                (
+                    Some((collator, total_paid)),
+                    <T as Config>::WeightInfo::pay_one_collator_reward(num_delegators as u32)
+                        + extra_weight,
```

Comment:
saturating_add

Reply:
weights are capped by max block weight which is much lower than u64::MAX